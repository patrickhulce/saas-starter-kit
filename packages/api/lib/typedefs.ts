import 'express' // tslint:disable-line

declare module 'express-serve-static-core' {
  /* tslint:disable */
  export interface Response {
    body?: any
  }
  /* tslint:enable */
}

export {}
