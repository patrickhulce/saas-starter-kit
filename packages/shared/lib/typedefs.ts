import {ISQLOptions} from 'klay'

export enum ModelID {
  Account = 'account',
  User = 'user',
}

export interface IAccount {
  id?: number
  name: string
  plan: AccountPlan
  createdAt?: Date
  updatedAt?: Date
}

export interface IUser {
  id?: number
  accountId: number
  role: AuthRole
  email: string
  password: string
  firstName: string
  lastName: string
  isVerified: boolean
  verificationKey?: string
  createdAt?: Date
  updatedAt?: Date
}

export enum AuthRole {
  Root = 'root',
  Admin = 'admin',
  User = 'user',
  Anonymous = 'anonymous',
}

export enum Permission {
  RootAccess = '_root_',
  AccountManage = 'account:manage',
  AccountView = 'account:view',
  UserPassword = 'user:password',
  UserProfile = 'user:profile',
  UserView = 'user:view',
}

export enum AccountPlan {
  Free = 'free',
  Bronze = 'bronze',
  Silver = 'silver',
  Gold = 'gold',
}

export interface IConf {
  debug: boolean
  production: boolean
  isUnderTest: boolean

  origin: string
  cdnAppURL: string
  termsOfServiceURL: string
  displayName: string

  secret: string
  database: ISQLOptions

  mailslurp: {apiKey: string}
  sparkpost: {fromAddress: string; apiKey: string; sendToSink: boolean}
}
