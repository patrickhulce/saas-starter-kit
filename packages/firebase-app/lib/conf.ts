import * as functions from 'firebase-functions'
import {IConf} from '../../shared/lib/typedefs'
import conf from '../../shared/lib/conf'

conf.production = !!functions.config().production
conf.apiPathPrefix = '/api'
conf.database.connectionURL = functions.config().sqlURL

export default conf
