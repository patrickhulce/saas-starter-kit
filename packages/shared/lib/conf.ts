import {IConf, JsEnvironment} from './typedefs'

const conf: IConf = {
  debug: true,
  production: false,
  isUnderTest: typeof it === 'function', // tslint:disable-line

  domain: 'THE_DOMAIN',
  displayName: 'THE_PRODUCT_DISPLAY_NAME',
  apiPathPrefix: '/api',

  jsEnvironment: JsEnvironment.Server,
  secret: process.env.SECRET || 'super-secret-string',
  database: {
    connectionURL: process.env.APP_MYSQL_URL || 'mysql://root@localhost/the_product',
  },

  mailslurp: {
    apiKey: process.env.MAILSLURP_API_KEY || '',
  },

  sparkpost: {
    sendToSink: true,
    apiKey: process.env.SPARKPOST_API_KEY || '',
    fromAddress: process.env.SPARKPOST_FROM || '',
  },
}

if (process.env.APP_ENV === 'development') {
  conf.sparkpost.sendToSink = false
}

if (process.env.APP_ENV === 'firebase-development') {
  conf.sparkpost.sendToSink = false
}

if (process.env.APP_ENV === 'firebase-production') {
  conf.debug = false
  conf.production = true
  conf.sparkpost.sendToSink = false
}

if (!conf.sparkpost.fromAddress || conf.isUnderTest) {
  conf.sparkpost.fromAddress = `noreply@${conf.domain}`
}

// tslint:disable-next-line
export default conf
