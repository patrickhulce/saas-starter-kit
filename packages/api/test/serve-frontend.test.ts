import conf from '../../shared/lib/conf'

import {IState} from './typedefs'

describe('serve frontend', () => {
  const state: IState = {}

  require('./steps/initialize.test')(state)
  require('./steps/account-setup.test')(state)

  describe('serve frontend', () => {
    let fetchResponse, fetchTextFn: jest.Mock

    beforeEach(() => {
      // @ts-ignore
      global.__lruCache = {}
      conf.stripe.publicKey = '<key>'
      fetchTextFn = jest.fn()
      fetchResponse = {status: 200, text: fetchTextFn}

      __fetch.mockReset()
      __fetch.mockResolvedValue(fetchResponse)
    })

    it('should serve the login page', async () => {
      fetchTextFn.mockResolvedValueOnce('<hash>').mockResolvedValueOnce('<document>')

      const response = await fetch(`${state.baseURL}/login`)
      expect(response.status).toBe(200)
      const text = await response.text()
      expect(text).toEqual('<document>')
      expect(__fetch.mock.calls[1][0]).toContain('<hash>/login.html')
    })

    it('should cache subsequent fetches for same hash', async () => {
      fetchTextFn
        .mockResolvedValueOnce('<hash>')
        .mockResolvedValueOnce('<document>')
        .mockResolvedValueOnce('<hash>')
        .mockRejectedValue(new Error('should have been cached'))

      await fetch(`${state.baseURL}/login`)

      const response = await fetch(`${state.baseURL}/login`)
      expect(response.status).toBe(200)
      const text = await response.text()
      expect(text).toEqual('<document>')
      expect(__fetch).toHaveBeenCalledTimes(3)
    })

    it('should not return from cache for new hash', async () => {
      fetchTextFn
        .mockResolvedValueOnce('<hash>')
        .mockResolvedValueOnce('<document>')
        .mockResolvedValueOnce('<hash1>')
        .mockResolvedValueOnce('NEW')

      await fetch(`${state.baseURL}/login`)

      const response = await fetch(`${state.baseURL}/login`)
      expect(response.status).toBe(200)
      const text = await response.text()
      expect(text).toEqual('NEW')
      expect(__fetch).toHaveBeenCalledTimes(4)
    })

    it('should redirect to login page when no cookies set', async () => {
      fetchTextFn.mockResolvedValueOnce('<hash>').mockResolvedValueOnce('<document>')

      const response = await fetch(state.baseURL)
      expect(response.status).toBe(200)
      const text = await response.text()
      expect(text).toEqual('<document>')
      expect(__fetch.mock.calls[1][0]).toContain('<hash>/login.html')
    })

    it('should serve the app page', async () => {
      fetchTextFn.mockResolvedValueOnce('<hash>').mockResolvedValueOnce('<document>')

      const response = await fetch(state.baseURL, {
        headers: {cookie: `token=${state.token}`},
      })

      expect(response.status).toBe(200)
      const text = await response.text()
      expect(text).toEqual('<document>')
      expect(__fetch.mock.calls[1][0]).toContain('<hash>/index.html')
    })

    it('should serve the account page', async () => {
      fetchTextFn.mockResolvedValueOnce('<hash>').mockResolvedValueOnce('<document>')

      const response = await fetch(`${state.baseURL}/account`, {
        headers: {cookie: `token=${state.token}`},
      })

      expect(response.status).toBe(200)
      const text = await response.text()
      expect(text).toEqual('<document>')
      expect(__fetch.mock.calls[1][0]).toContain('<hash>/account.html')
    })

    it('should inject stripe public key', async () => {
      fetchTextFn.mockResolvedValueOnce('<hash1>').mockResolvedValueOnce('<html><head>')

      const response = await fetch(`${state.baseURL}/account`, {
        headers: {cookie: `token=${state.token}`},
      })

      expect(response.status).toBe(200)
      const text = await response.text()
      expect(text).toMatchSnapshot()
    })
  })

  require('./steps/teardown.test')(state)
})
