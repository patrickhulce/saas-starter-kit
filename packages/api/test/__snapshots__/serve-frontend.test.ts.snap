// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serve frontend serve frontend should inject stripe public key 1`] = `"<html><head><script>window.STRIPE_PUBLIC_KEY = \\"<key>\\"</script>"`;
